include/runtime.h

```diff
@@ -44,45 +35,41 @@ namespace detail {
 		runtime& operator=(const runtime&) = delete;
 		runtime& operator=(runtime&&) = delete;
 
-		~runtime();
+		virtual ~runtime() = default;
 
-		template <typename CGF>
-		task_id submit(CGF&& cgf) {
-			// TODO move task-node generation out of handler to avoid task_manager header dependency here
-			require_call_from_application_thread();
-			maybe_prune_task_graph();
-			return m_task_mngr->submit_command_group(std::forward<CGF>(cgf));
-		}
+		virtual task_id submit(raw_command_group&& cg) = 0;
+
+		virtual task_id fence(buffer_access access, std::unique_ptr<task_promise> fence_promise) = 0;
 
-		task_id fence(buffer_access_map access_map, side_effect_map side_effects, std::unique_ptr<task_promise> fence_promise);
+		virtual task_id fence(host_object_effect effect, std::unique_ptr<task_promise> fence_promise) = 0;
 
-		task_id sync(detail::epoch_action action);
+		virtual task_id sync(detail::epoch_action action) = 0;
 
-		void create_queue();
+		virtual void create_queue() = 0;
 
-		void destroy_queue();
+		virtual void destroy_queue() = 0;
 
-		allocation_id create_user_allocation(void* ptr);
+		virtual allocation_id create_user_allocation(void* ptr) = 0;
 
-		buffer_id create_buffer(const range<3>& range, size_t elem_size, size_t elem_align, allocation_id user_aid);
+		virtual buffer_id create_buffer(const range<3>& range, size_t elem_size, size_t elem_align, allocation_id user_aid) = 0;
 
-		void set_buffer_debug_name(buffer_id bid, const std::string& debug_name);
+		virtual void set_buffer_debug_name(buffer_id bid, const std::string& debug_name) = 0;
 
-		void destroy_buffer(buffer_id bid);
+		virtual void destroy_buffer(buffer_id bid) = 0;
 
-		host_object_id create_host_object(std::unique_ptr<host_object_instance> instance /* optional */);
+		virtual host_object_id create_host_object(std::unique_ptr<host_object_instance> instance /* optional */) = 0;
 
-		void destroy_host_object(host_object_id hoid);
+		virtual void destroy_host_object(host_object_id hoid) = 0;
 
-		reduction_id create_reduction(std::unique_ptr<reducer> reducer);
+		virtual reduction_id create_reduction(std::unique_ptr<reducer> reducer) = 0;
 
-		bool is_dry_run() const { return m_cfg->is_dry_run(); }
+		virtual bool is_dry_run() const = 0;
 
-		void set_scheduler_lookahead(experimental::lookahead lookahead);
+		virtual void set_scheduler_lookahead(experimental::lookahead lookahead) = 0;
 
-		void flush_scheduler();
+		virtual void flush_scheduler() = 0;
```

Comment:
As discussed offline, it would be nice to find a solution here that does not require us to make these public interfaces virtual.